### PR TITLE
Bug/select wallet incomplete  / Refactor wallet Services

### PR DESCRIPTION
**Pull request by cmgustavo**
* Fixes when selecting an incomplete wallet
* Refactor of txService to walletService

### CODE REVIEW

File: src/js/services/walletService.js [hunk 1]
```diff
@@ -0,0 +1,163 @@
+'use strict';
+
+angular.module('copayApp.services').factory('walletService', function($log, lodash, trezor, ledger, storageService) {
+  var root = {};
+
+  var _signWithLedger = function(txp, client, cb) {
```
Comment (matiu) [inline src/js/services/walletService.js]: client should be first  para, then txp

File: src/js/services/walletService.js [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        return cb(result.message || result.error);
+
+      txp.signatures = lodash.map(result.signatures, function(s) {
+        return s.substring(0, s.length - 2);
+      });
+      return client.signTxProposal(txp, cb);
+    });
+  };
+
+  var _signWithTrezor = function(txp, client, cb) {
```
Comment (matiu) [inline src/js/services/walletService.js]: client should be first  para, then txp

File: src/js/services/walletService.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    if(!client.isComplete()) 
+      return cb('WALLET_NOT_COMPLETE');
+    root.isBackupNeeded(client, function(needsBackup) {
+      if (needsBackup)
+        return cb('WALLET_NEEDS_BACKUP');
+      return cb();
+    });
+  };
+
+  root.createTx = function(txp, client, cb) {
```
Comment (matiu) [inline src/js/services/walletService.js]: client should be first  para, then txp 

(same in the others)

File: src/js/controllers/backup.js [hunk 4]
```diff
@@ -46,7 +46,7 @@ angular.module('copayApp.controllers').controller('backupController',
     function initWords() {
       var words = fc.getMnemonic();
       self.xPrivKey = fc.credentials.xPrivKey;
-      profileService.lockFC();
+      profileService.lockFC(fc);
```
Comment (matiu) [inline src/js/controllers/backup.js]: function should be lock(fc)

File: src/js/controllers/backup.js [hunk 5]
```diff
@@ -74,7 +74,7 @@ angular.module('copayApp.controllers').controller('backupController',
             $scope.$apply();
           }, 1);
 
-          profileService.unlockFC({}, function(err) {
+          profileService.unlockFC(fc, function(err) {
```
Comment (matiu) [inline src/js/controllers/backup.js]: `unlock(fc, ...`

File: src/js/controllers/buyGlidera.js [hunk 6]
```diff
@@ -50,14 +50,17 @@ angular.module('copayApp.controllers').controller('buyGlideraController',
...
-          }
-          $modalInstance.close({
-            'walletId': walletId, 
-            'walletName': walletName, 
+          var client = profileService.getClient(walletId);
+          walletService.isReady(client, function(err) {
+            if (err) {
+              self.error = err;
+              $modalInstance.dismiss('cancel');
+            } else {
```
Comment (matiu) [inline src/js/controllers/buyGlidera.js]: dont use `else` in `if (err)`; use return

File: src/js/controllers/sellCoinbase.js [hunk 7]
```diff
@@ -202,28 +205,29 @@ angular.module('copayApp.controllers').controller('sellCoinbaseController',
...
             outputs: outputs,
             message: comment,
-            payProUrl: null
+            payProUrl: null,
+            excludeUnconfirmedUtxos: configWallet.spendUnconfirmed ? false : true,
+            feeLevel: walletSettings.feeLevel || 'normal'
           };
 
-          txService.createTx(opts, function(err, txp) {
+          walletService.createTx(txp, fc, function(err, createdTxp) {
```
Comment (matiu) [inline src/js/controllers/sellCoinbase.js]: args should be (client, txp, ...)

**Comment by matiu** [comment]
great work!

will continue to review tomorrow.

File: src/js/controllers/sellGlidera.js [hunk 8]
```diff
@@ -127,87 +126,112 @@ angular.module('copayApp.controllers').controller('sellGlideraController',
...
+                          };
+                          self.loading = 'Selling bitcoin...';
+                          glideraService.sell(token, twoFaCode, data, function(err, data) {
+                            self.loading = null;
+                            if (err) {
+                              self.error = err.message || bwsError.msg(err);
+                              $timeout(function() {
+                                $scope.$emit('Local/GlideraError');
+                              }, 100);
+                            } else {
```
Comment (matiu) [inline src/js/controllers/sellGlidera.js]: please use `return` in `if(err)` and dont use `else`

File: src/js/controllers/walletHome.js [hunk 9]
```diff
@@ -251,28 +251,30 @@ angular.module('copayApp.controllers').controller('walletHomeController', functi
...
-
-          $scope.gettingAddress = true;
-          $scope.selectedWalletName = walletName;
-          $timeout(function() {
-            $scope.$apply();
-          });
+        var client = profileService.getClient(walletId);
+        $scope.errorSelectedWallet = {};
+
+        walletService.isReady(client, function(err) {
```
Comment (matiu) [inline src/js/controllers/walletHome.js]: Code looks better this way! 👍

File: src/js/controllers/walletHome.js [hunk 10]
```diff
@@ -413,25 +415,49 @@ angular.module('copayApp.controllers').controller('walletHomeController', functi
...
             $scope.error = err;
-            $timeout(function() {
-              $scope.$digest();
-            });
             return;
           }
-          $modalInstance.close(txp);
-          return;
+
+          profileService.unlockFC(client, function(err) {
```
Comment (matiu) [inline src/js/controllers/walletHome.js]: lock/unlock should be on `walletService`

File: src/js/services/fingerprintService.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+          return cb(gettextCatalog.getString('Touch ID Failed') + ': ' + msg.localizedDescription);
+        }
+      );
+    } catch (e) {
+      $log.debug('Touch ID Failed:' + JSON.stringify(e));
+      return cb(gettextCatalog.getString('Touch ID Failed'));
+    };
+  };
+
+  root.check = function(client, config, cb) {
```
Comment (matiu) [inline src/js/services/fingerprintService.js]: If we agree that configService is OK to be shared in services, we should not pass `config`

**Comment by matiu** [comment]
review done. I will start doing some manual test

File: src/js/controllers/walletHome.js [hunk 12]
```diff
@@ -458,11 +492,12 @@ angular.module('copayApp.controllers').controller('walletHomeController', functi
 
 
       $scope.remove = function(txp) {
+        var client = profileService.focusedClient;
         self.setOngoingProcess(gettextCatalog.getString('Deleting payment'));
         $scope.loading = true;
         $scope.error = null;
         $timeout(function() {
-          fc.removeTxProposal(txp, function(err, txpb) {
+          walletService.removeTx(client, txp, function(err) {}
```
Comment (matiu) [inline src/js/controllers/walletHome.js]: extra `}` is a syntax error

**Comment by cmgustavo** [comment]
Updated

**Comment by matiu** [comment]
Great job!

**Pull request merged**